Lib/Src/CSoftPwmOutput.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * CSoftPwmOutput.cpp
+ *
+ *  Created on: 20 Feb 2022
+ *      Author: salavat.magazov
+ */
+
+#include "CSoftPwmOutput.h"
+
+TIM_HandleTypeDef *CSoftPwmOutput::sp_timer = nullptr;
+uint32_t CSoftPwmOutput::s_period = 100;
+uint32_t CSoftPwmOutput::s_counter = 0;
+CSoftPwmOutput *CSoftPwmOutput::sp_first_instance = nullptr;
+
+/**
+ * @brief Construct an instance of soft PWM timer based on GPIO.
+ *
+ * @param p_gpio pointer to GPIO to be controlled by soft PWM.
+ */
+CSoftPwmOutput::CSoftPwmOutput(GPIO_TypeDef *p_port, uint16_t pin)
+    : CGpioWrapper(p_port, pin),
+      mp_next_instance(nullptr),
+      m_duty_cycle(0)
+{
+    if (sp_first_instance == nullptr)
+    {
+        sp_first_instance = this;
+    }
+    else
+    {
+        CSoftPwmOutput *p_previous_instance = getLastInstance();
+        p_previous_instance->mp_next_instance = this;
+    }
+    mp_next_instance = nullptr;
+    CGpioWrapper::set(false);
+}
+
+/**
+ * @brief Set duty cycle of the PWM output.
+ *
+ * @param duty_cycle_percent Percentage value of duty cycle.
+ */
+void CSoftPwmOutput::setDutyCycle(float duty_cycle_percent)
+{
+    duty_cycle_percent *= s_period + 1;
+    m_duty_cycle = duty_cycle_percent / 100;
+    if (m_duty_cycle > s_period)
+    {
+        m_duty_cycle = s_period;
+    }
+}
+
+/**
+ * @brief Get current duty cycle.
+ *
+ * @return Current duty cycle generated by the class.
+ */
+float CSoftPwmOutput::getDutyCycle() const
+{
+    return (m_duty_cycle / s_period);
+}
+
+/**
+ * @brief Initialise the class.
+ * @note This init only needs to be called once for all instances.
+ * For operation with timer, it is assumed that timer has been fully setup.
+ * For operation without a timer, user needs to ensure repeated calls to tick()
+ * method to keep the PWM operational.
+ *
+ * @param period Period of the PWM.
```

Comment:
Is the period in ms?

Reply:
No. It's in raw units. Exact timer depends on how the timer is setup. If you setup timer to update every 1ms and set period to be 100, then it's going to be 100ms. If timer updates every 2ms, then period suddenly doubles.
Perhaps worth putting that in doxygen note.